[nnc] Added implementation for aten::sign op

navahgar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #64205
* __->__ #64033

Differential Revision: [D30579197](https://our.internmc.facebook.com/intern/diff/D30579197)

navahgar committed:
[nnc] Added sign op

[ghstack-poisoned]

bertmaher commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -2056,6 +2057,18 @@ Tensor tensorexpr::computeOperandValue(
+          outputType,
+          [](const ExprHandle& a) {
+            return CompareSelect::make(
+                a, ExprHandle(0), ExprHandle(-1), ExprHandle(1), kLT);
```
I was expecting that this would be incorrect for -0.0 but...
```
>>> torch.tensor([-0.0, 0.0]).sign()
tensor([0., 0.])
```

Seems fishy.

bertmaher replied:
Oh wait, I see what's going on.  torch.sign returns 0 for +/-0.0: https://pytorch.org/docs/stable/generated/torch.sign.html.  So this expression is wrong for either flavor of 0 :-)

navahgar replied:
Thanks for pointing those out. Handled all those cases now.

bertmaher reviewed (changes requested):
Could you add test cases please?  Make sure to cover +/- 0.0, inf, and nan

navahgar committed:
Update on "[nnc] Added implementation for aten::sign op"


Differential Revision: [D30579197](https://our.internmc.facebook.com/intern/diff/D30579197)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30579197).

navahgar commented:
Actually this PR was WIP.. I didn't realize that I had sent it out for review.. Anyways, thanks for pointing those out. 

> Could you add test cases please? Make sure to cover +/- 0.0, inf, and nan

Added test for all the cases involved.

navahgar committed:
Update on "[nnc] Added implementation for aten::sign op"


Differential Revision: [D30579197](https://our.internmc.facebook.com/intern/diff/D30579197)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30579197).

navahgar committed:
Update on "[nnc] Added implementation for aten::sign op"


Differential Revision: [D30579197](https://our.internmc.facebook.com/intern/diff/D30579197)

[ghstack-poisoned]

navahgar commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -505,6 +505,24 @@ void promoteInputs(std::vector<ExprHandle>& inputs, const int typeConstraints) {
   }
 }
 
+ExprHandle promoteIntegerToDefaultType(const ExprHandle& e) {
```
This code was just moved from below to define it in appropriate namespace since this is now exposed through the header and used in other files.

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30579197).

navahgar committed:
Update on "[nnc] Added implementation for aten::sign op"


Differential Revision: [D30579197](https://our.internmc.facebook.com/intern/diff/D30579197)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30579197).

navahgar committed:
Update on "[nnc] Added implementation for aten::sign op"


Differential Revision: [D30579197](https://our.internmc.facebook.com/intern/diff/D30579197)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30579197).

navahgar committed:
Update on "[nnc] Added implementation for aten::sign op"


Differential Revision: [D30579197](https://our.internmc.facebook.com/intern/diff/D30579197)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30579197).

bertmaher commented on torch/csrc/jit/tensorexpr/operators/unary.cpp:
```diff
@@ -0,0 +1,39 @@
+        // (isnan(inp) == 1) ? 0.0f : v2
+        auto res = CompareSelect::make(
+            isnan(inp), ExprHandle(1), ExprHandle(0.0f), v2, kEQ);
+        // Final expression:
+        //    (isnan(inp) == 1) ?
+        //        0.0f :
+        //        (inp == 0.0f) ?
+        //            0.0f :
+        //            (inp < 0.0f) ? -1.0f : 1.0f
+        return res;
+      });
```
A couple issues here:
1. I'd probably use the same expression as ATen `(0<a) - (a<0)`, https://github.com/pytorch/pytorch/blob/a8ea3461145551bb85855135eafb84fc754fd320/aten/src/ATen/native/cpu/UnaryOpsKernel.cpp#L299
2. I'm not sure why this doesn't fail in unit tests, but the dtypes in the `CompareSelects` aren't going to be right since you're explicitly creating floats and comparing to whatever the input dtype is (so, double would probably break).  Instead of `ExprHandle(-1.0f)` you should probably do `immLike(inp, -1.0f)`.

navahgar replied:
1. Changed the implementation to the one in ATen (I had initially thought of doing that as a follow up..)

2. The unit test wasn't failing because it only had `float` inputs. I updated the test to add `double` as well, and fixed the implementation accordingly.

navahgar committed:
Update on "[nnc] Added implementation for aten::sign op"


Differential Revision: [D30579197](https://our.internmc.facebook.com/intern/diff/D30579197)

[ghstack-poisoned]

navahgar committed:
Update on "[nnc] Added implementation for aten::sign op"


Differential Revision: [D30579197](https://our.internmc.facebook.com/intern/diff/D30579197)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30579197).

navahgar commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -41,7 +41,20 @@ static bool checkTypes(const ScalarType highType, const int typeConstraints) {
+  return s + " " + generic_error_message;
+}
+
+ExprHandle promoteToDtype(ExprHandle e, ScalarType dt) {
```
This change is only to expose `promoteToDtype` to other files.

navahgar committed:
Update on "[nnc] Added implementation for aten::sign op"


Differential Revision: [D30579197](https://our.internmc.facebook.com/intern/diff/D30579197)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30579197).

navahgar committed:
Update on "[nnc] Added implementation for aten::sign op"


Differential Revision: [D30579197](https://our.internmc.facebook.com/intern/diff/D30579197)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30579197).

navahgar committed:
Update on "[nnc] Added implementation for aten::sign op"


Differential Revision: [D30579197](https://our.internmc.facebook.com/intern/diff/D30579197)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30579197).

navahgar committed:
Update on "[nnc] Added implementation for aten::sign op"


Differential Revision: [D30579197](https://our.internmc.facebook.com/intern/diff/D30579197)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30579197).
